Can't put custom dictionary in xaml resource

### Current behavior

Compilation error. Seems like a code-gen issue in non-UWP platforms.
```
1>C:\Users\Peter\Desktop\test\UnoApp1\UnoApp1.Mobile\Android\AndroidManifest.xml : warning XA4216: AndroidManifest.xml //uses-sdk/@android:minSdkVersion '16' is less than API-19, this configuration is not supported.
1>Skipping analyzers to speed up the build. You can execute 'Build' or 'Rebuild' command to run analyzers.
1>C:\Users\Peter\Desktop\test\UnoApp1\UnoApp1.Mobile\Uno.UI.SourceGenerators\Uno.UI.SourceGenerators.XamlGenerator.XamlCodeGenerator\App_1e3e359cf1b3eb7940fb0feeae07a8f1.cs(99,8,99,11): error CS7036: There is no argument given that corresponds to the required parameter 'value' of 'Dictionary<string, Style>.Add(string, Style)'
1>Done building project "UnoApp1.Mobile.csproj" -- FAILED.
```

### Expected behavior

Compile successfully.

### How to reproduce it (as minimally and precisely as possible)

1. Create new project, using `.NET6 UWP` template
2. Add a new class, call `Dict.cs` in Shared project
```cs
using System;
using System.Collections.Generic;
using System.Text;
using Windows.UI.Xaml;

namespace UnoApp1
{
    public class StringToStyleDictionary : Dictionary<string, Style>
    {
    }
}
```
3. Add an entry in `App.xaml`
```
<Application
    x:Class="UnoApp1.App"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:UnoApp1">
	
    <Application.Resources>
        <ResourceDictionary>
            <ResourceDictionary.MergedDictionaries>
                <XamlControlsResources xmlns="using:Microsoft.UI.Xaml.Controls" />
                <!-- Add resource dictionaries here -->                
            </ResourceDictionary.MergedDictionaries>
            <local:StringToStyleDictionary x:Key="MyResource">
                <Style x:Key="MyStyle" TargetType="Button">

                </Style>
            </local:StringToStyleDictionary>
        </ResourceDictionary>
    </Application.Resources>

</Application>
```
4. Build and run for Android, or anything other than `UWP`
![image](https://github.com/unoplatform/uno/assets/42881734/9a26c6c9-02ee-462a-9412-5139df933991)

This same code runs in UWP
![image](https://github.com/unoplatform/uno/assets/42881734/b995f22a-c16d-4135-be40-eddb7923bf12)


### Workaround

No.

### Works on UWP/WinUI

Yes

### Environment

Uno.SourceGenerationTasks

### NuGet package version(s)

_No response_

### Affected platforms

WebAssembly, Android, iOS, macOS (AppKit), Mac Catalyst, Skia (WPF), Skia (GTK on Linux/macOS/Windows), Skia (Linux Framebuffer), Skia (Tizen)

### IDE

Visual Studio 2022

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

--- a/src/SamplesApp/SamplesApp.Shared/App.xaml
+++ b/src/SamplesApp/SamplesApp.Shared/App.xaml
@@ -4,6 +4,7 @@
 			 xmlns:xamarin="http://nventive.com/xamarin"
 			 xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
 			 xmlns:ios="http://nventive.com/ios"
+			 xmlns:local="using:SamplesApp"
 			 xmlns:toolkit="using:Uno.UI.Toolkit"
 			 mc:Ignorable="ios"
 			 RequestedTheme="Light">
@@ -32,6 +33,12 @@
 				<ResourceDictionary Source="Windows_UI_Xaml_Controls\ListView\MeasureDetectorControl\MeasureDetectorControl.xaml" />
 				<MapResources xmlns="using:Uno.UI.Maps"/>
 			</ResourceDictionary.MergedDictionaries>
+
+			 <local:StringToStyleDictionary x:Key="MyResource">
+                <Style x:Key="MyStyle" TargetType="Button">
+
+                </Style>
+            </local:StringToStyleDictionary>
 
 			<SolidColorBrush x:Key="GlobalStaticResource_Test01"
 							 Color="Purple" />

--- a/src/SamplesApp/SamplesApp.Shared/SamplesApp.Shared.projitems
+++ b/src/SamplesApp/SamplesApp.Shared/SamplesApp.Shared.projitems
@@ -27,6 +27,7 @@
     <Compile Include="$(MSBuildThisFileDirectory)Samples\UnitTests\UnitTestsPage.xaml.cs">
       <DependentUpon>UnitTestsPage.xaml</DependentUpon>
     </Compile>
+    <Compile Include="$(MSBuildThisFileDirectory)StringToStyleDictionary.cs" />
   </ItemGroup>
   <ItemGroup>
     <Page Include="$(MSBuildThisFileDirectory)MainPage.xaml">

--- a/src/SamplesApp/SamplesApp.Shared/StringToStyleDictionary.cs
+++ b/src/SamplesApp/SamplesApp.Shared/StringToStyleDictionary.cs
@@ -1,0 +1,8 @@
+﻿using System.Collections.Generic;
+using Windows.UI.Xaml;
+
+namespace SamplesApp;
+
+class StringToStyleDictionary : Dictionary<string, Style>
+{
+}

--- a/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs
+++ b/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs
@@ -3313,6 +3313,11 @@ namespace Uno.UI.SourceGenerators.XamlGenerator
 									writer.AppendLineIndented($"{closureName}.Add(");
 									using (writer.Indent())
 									{
+										if (objectDefinitionType?.AllInterfaces.Any(i => i.OriginalDefinition.Equals(Generation.IDictionaryOfTKeySymbol.Value, SymbolEqualityComparer.Default)) == true &&
+											GetDictionaryResourceKey(item) is string dictionaryKey)
+										{
+											writer.AppendLineIndented($"\"{dictionaryKey}\",");
+										}
 										BuildChild(writer, member, item);
 									}
 									writer.AppendLineIndented(");");

